From a38fe3a50427bd0ffdd86d0a565371d992027438 Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?R=C3=A9mi=20Bernon?= <rbernon@codeweavers.com>
Date: Thu, 29 May 2025 10:44:33 +0200
Subject: [PATCH] frames/window-tracker: Move frame existence check inside of
 set_up_frame

Avoiding creating frames multiple times when listen_set_up_frame is
called from a ConfigureNotify event after the frame has already been
created.

Fixes: 4d194b8e1 ("frames/window-tracker: Handle override-redirect attribute change.")
Closes: https://gitlab.gnome.org/GNOME/mutter/-/issues/4138
Part-of: <https://gitlab.gnome.org/GNOME/mutter/-/merge_requests/4465>
(cherry picked from commit a1d6e5771bbd704c65c8079310102637a5aab071)
---
 src/frames/meta-window-tracker.c | 9 ++++++---
 1 file changed, 6 insertions(+), 3 deletions(-)

diff --git a/src/frames/meta-window-tracker.c b/src/frames/meta-window-tracker.c
index 247248d108..18801f5c3f 100644
--- a/src/frames/meta-window-tracker.c
+++ b/src/frames/meta-window-tracker.c
@@ -125,6 +125,11 @@ set_up_frame (MetaWindowTracker *window_tracker,
   unsigned long data[1];
   GtkWidget *frame;
 
+  frame = g_hash_table_lookup (window_tracker->client_windows,
+                               GUINT_TO_POINTER (xwindow));
+  if (frame)
+    return;
+
   /* Double check it's not a request for a frame of our own. */
   if (g_hash_table_contains (window_tracker->frames,
                              GUINT_TO_POINTER (xwindow)))
@@ -287,9 +292,7 @@ on_xevent (GdkDisplay *display,
            xevent->xproperty.atom ==
            gdk_x11_get_xatom_by_name_for_display (display, "_MUTTER_NEEDS_FRAME"))
     {
-      if (xevent->xproperty.state == PropertyNewValue &&
-          !g_hash_table_contains (window_tracker->client_windows,
-                                  GUINT_TO_POINTER (xwindow)))
+      if (xevent->xproperty.state == PropertyNewValue)
         set_up_frame (window_tracker, xwindow);
       else if (xevent->xproperty.state == PropertyDelete)
         remove_frame (window_tracker, xwindow);
-- 
GitLab

